File: Core/src/main/java/com/cyrillrx/android/toolbox/logger/Logger.java
package com.cyrillrx.android.toolbox.logger;

import android.annotation.SuppressLint;
import android.content.Context;
import android.view.Gravity;
import android.widget.Toast;

import java.util.HashSet;
import java.util.Set;

/**
 * This class wraps instances of the {@link LogChild} interface.<br />
 * It allows to customize the logging conditions.
 *
 * @author Cyril Leroux
 *         Created on 03/09/12
 */
@SuppressWarnings("unused")
public class Logger {

    private static final String ERROR_ALREADY_INITIALIZED = "initialize() has already been called.";
    private static final String ERROR_INITIALIZE_FIRST = "Call initialize() before using the Logger.";
    private static Logger sInstance;

    private final Set<LogChild> mLoggers;

    private Toast mDebugToast;

    /**
     * @param context The application context to initialize the debug toast or null.
     */
    @SuppressLint("ShowToast")
    private Logger(Context context) {

        mLoggers = new HashSet<>();

        if (context != null) {
            mDebugToast = Toast.makeText(context, "", Toast.LENGTH_LONG);
            mDebugToast.setGravity(Gravity.TOP, 0, 50);
        }
    }

    /**
     * Initializes the Logger.<br />
     * Provides a context to enable a simple Toast used to display debug messages to the developer.
     *
     * @param context The application context to initialize the debug toast or null.
     */
    public static void initialize(Context context) {
        throwIfInitialized();

        sInstance = new Logger(context);
    }

    /**
     * Initializes the Logger without the debug logger.
     */
    public static void initialize() {
        throwIfInitialized();

        sInstance = new Logger(null);
    }

    /**
     * Shows a debug toast.
     *
     * @param message The message to toast.
     */
    public static synchronized void toast(String message) {
        throwIfNotInitialized();

        final Toast toast = sInstance.mDebugToast;
        if (toast == null) { return; }

        toast.setText(message);
        toast.show();
    }

    public static synchronized void addChild(LogChild child) {
        throwIfNotInitialized();

        sInstance.mLoggers.add(child);
    }

    public static synchronized void removeChild(LogChild child) {
        throwIfNotInitialized();

        sInstance.mLoggers.remove(child);
    }

    public static synchronized void verbose(String tag, String message) {
        throwIfNotInitialized();

        for (LogChild log : sInstance.mLoggers) {
            log.verbose(tag, message);
        }
    }

    public static synchronized void verbose(String tag, String message, Throwable throwable) {
        throwIfNotInitialized();

        for (LogChild log : sInstance.mLoggers) {
            log.verbose(tag, message, throwable);
        }
    }

    public static synchronized void debug(String tag, String message) {
        throwIfNotInitialized();

        for (LogChild log : sInstance.mLoggers) {
            log.debug(tag, message);
        }
    }

    public static synchronized void debug(String tag, String message, Throwable throwable) {
        throwIfNotInitialized();

        for (LogChild log : sInstance.mLoggers) {
            log.debug(tag, message, throwable);
        }
    }

    public static synchronized void info(String tag, String message) {
        throwIfNotInitialized();

        for (LogChild log : sInstance.mLoggers) {
            log.info(tag, message);
        }
    }

    public static synchronized void info(String tag, String message, Throwable throwable) {
        throwIfNotInitialized();

        for (LogChild log : sInstance.mLoggers) {
            log.info(tag, message, throwable);
        }
    }

    public static synchronized void warning(String tag, String message) {
        throwIfNotInitialized();

        for (LogChild log : sInstance.mLoggers) {
            log.warning(tag, message);
        }
    }

    public static synchronized void warning(String tag, String message, Throwable throwable) {
        throwIfNotInitialized();

        for (LogChild log : sInstance.mLoggers) {
            log.warning(tag, message, throwable);
        }
    }

    public static synchronized void error(String tag, String message) {
        throwIfNotInitialized();

        for (LogChild log : sInstance.mLoggers) {
            log.error(tag, message);
        }
    }

    public static synchronized void error(String tag, String message, Throwable throwable) {
        throwIfNotInitialized();

        for (LogChild log : sInstance.mLoggers) {
            log.error(tag, message, throwable);
        }
    }

    private static void throwIfInitialized() {
        if (sInstance != null) {
            throw new IllegalStateException(ERROR_ALREADY_INITIALIZED);
        }
    }

    private static void throwIfNotInitialized() {
        if (sInstance == null) {
            throw new IllegalStateException(ERROR_INITIALIZE_FIRST);
        }
    }
}

File: shared/logger/src/commonMain/kotlin/com/cyrillrx/logger/Logger.kt
package com.cyrillrx.logger

import kotlin.experimental.ExperimentalObjCName
import kotlin.jvm.JvmStatic
import kotlin.native.ObjCName

/**
 * This class wraps instances of the {@link LogChild} interface.
 * It allows to customize the logging conditions.
 *
 * @author Cyril Leroux
 *         Created on 03/09/2012.
 */
@OptIn(ExperimentalObjCName::class)
@ObjCName("KMPLogger")
object Logger {
    private val loggers: MutableSet<LogChild> = HashSet()

    private var catcher: ExceptionCatcher? = null

    fun release() {
        loggers.clear()
    }

    fun setCatcher(catcher: ExceptionCatcher) {
        this.catcher = catcher
    }

    @JvmStatic
    fun addChild(child: LogChild) {
        loggers.add(child)
    }

    fun removeChild(child: LogChild) {
        loggers.remove(child)
    }

    fun log(severity: Severity, tag: String, message: String, throwable: Throwable? = null) {
        for (logger in loggers) {
            try {
                logger.log(severity, tag, message, throwable)
            } catch (t: Throwable) {
                try {
                    catcher?.catchException(t)
                } catch (ignored: Exception) {
                    // Prevent the catcher from throwing an exception
                }
            }
        }
    }

    @JvmStatic
    fun verbose(tag: String, message: String, throwable: Throwable? = null) {
        log(Severity.VERBOSE, tag, message, throwable)
    }

    @JvmStatic
    fun debug(tag: String, message: String, throwable: Throwable? = null) {
        log(Severity.DEBUG, tag, message, throwable)
    }

    @JvmStatic
    fun info(tag: String, message: String, throwable: Throwable? = null) {
        log(Severity.INFO, tag, message, throwable)
    }

    @JvmStatic
    fun warning(tag: String, message: String, throwable: Throwable? = null) {
        log(Severity.WARN, tag, message, throwable)
    }

    @JvmStatic
    fun error(tag: String, message: String, throwable: Throwable? = null) {
        log(Severity.ERROR, tag, message, throwable)
    }
}
